### ISSUE TITLE
Standardize metrics and metric groups across all RL agents (PPO, SAC, DDPG, etc.)

### ISSUE THREAD

**Issue by mr8bit** [TensorAeroSpace/TensorAeroSpace#177]
## Summary

Implement a unified metrics system across all reinforcement learning agents in TensorAeroSpace. This standardization will enable consistent experiment comparison, simplify integration with external logging tools, and improve the overall developer experience.

## Problem Statement

Currently, different RL agents (PPO, SAC, DDPG, A3C, etc.) in the framework:
- Use inconsistent naming conventions for similar metrics
- Lack a unified structure for metric groups
- Make cross-algorithm comparison difficult and error-prone
- Have no common interface for accessing and exporting metrics
- Require manual mapping when integrating with logging platforms

### 1. Standardized Metric Groups

Organize all metrics into logical groups with consistent prefixes:

| Group | Description | Example Metrics |
|-------|-------------|-----------------|
| `rollout/` | Episode collection statistics | rewards, lengths, success rates |
| `train/` | Training process metrics | learning rate, gradient norms, updates |
| `loss/` | Loss function values | policy loss, value loss, entropy loss |
| `eval/` | Evaluation metrics | eval rewards, eval success rate |
| `time/` | Performance and timing | FPS, elapsed time, total steps |

#### Rollout Metrics
- [ ] `rollout/ep_rew_mean` — Mean episode reward
- [ ] `rollout/ep_rew_std` — Episode reward standard deviation
- [ ] `rollout/ep_len_mean` — Mean episode length
- [ ] `rollout/success_rate` — Success rate (task-dependent)

#### Training Metrics
- [ ] `train/learning_rate` — Current learning rate
- [ ] `train/n_updates` — Total number of gradient updates
- [ ] `train/entropy` — Policy entropy (for stochastic policies)
- [ ] `train/explained_variance` — Value function explained variance
- [ ] `train/clip_fraction` — Clipping fraction (PPO-specific)

#### Loss Metrics (Algorithm-Specific)
- [ ] `loss/policy` or `loss/actor` — Policy/actor loss
- [ ] `loss/value` or `loss/critic` — Value/critic loss
- [ ] `loss/entropy` — Entropy bonus loss
- [ ] `loss/alpha` — Temperature coefficient loss (SAC)
- [ ] `loss/q1`, `loss/q2` — Q-function losses (SAC, TD3)

#### Time Metrics
- [ ] `time/fps` — Training frames per second
- [ ] `time/total_timesteps` — Cumulative environment steps
- [ ] `time/iterations` — Training iterations completed
- [ ] `time/time_elapsed` — Wall-clock time in seconds

### 3. Implementation: BaseMetricsLogger

Create a unified logging interface that all agents inherit from:

```python
from abc import ABC, abstractmethod
from typing import Dict, Any, Optional
from dataclasses import dataclass, field
from collections import defaultdict

@dataclass
class MetricGroup:
    """Container for a group of related metrics."""
    prefix: str
    metrics: Dict[str, float] = field(default_factory=dict)
    
    def add(self, name: str, value: float) -> None:
        self.metrics[f"{self.prefix}/{name}"] = value
    
    def to_dict(self) -> Dict[str, float]:
        return self.metrics.copy()


class BaseMetricsLogger(ABC):
    """
    Unified metrics logging interface for all RL agents.
    
    Supports multiple backends: TensorBoard, wandb, console, CSV.
    """
    
    def __init__(
        self,
        tensorboard_writer=None,
        use_wandb: bool = False,
        log_interval: int = 1,
        verbose: int = 1
    ):
        self.writer = tensorboard_writer
        self.use_wandb = use_wandb
        self.log_interval = log_interval
        self.verbose = verbose
        self._step = 0
        self._metrics_history = defaultdict(list)
    
    def log_rollout(
        self,
        ep_rew_mean: float,
        ep_len_mean: float,
        ep_rew_std: Optional[float] = None,
        success_rate: Optional[float] = None,
        **kwargs
    ) -> None:
        """Log episode rollout statistics."""
        metrics = MetricGroup("rollout")
        metrics.add("ep_rew_mean", ep_rew_mean)
        metrics.add("ep_len_mean", ep_len_mean)
        if ep_rew_std is not None:
            metrics.add("ep_rew_std", ep_rew_std)
        if success_rate is not None:
            metrics.add("success_rate", success_rate)
        for key, value in kwargs.items():
            metrics.add(key, value)
        self._log_metrics(metrics.to_dict())
    
    def log_training(
        self,
        learning_rate: float,
        n_updates: int,
        entropy: Optional[float] = None,
        explained_variance: Optional[float] = None,
        **kwargs
    ) -> None:
        """Log training process metrics."""
        metrics = MetricGroup("train")
        metrics.add("learning_rate", learning_rate)
        metrics.add("n_updates", n_updates)
        if entropy is not None:
            metrics.add("entropy", entropy)
        if explained_variance is not None:
            metrics.add("explained_variance", explained_variance)
        for key, value in kwargs.items():
            metrics.add(key, value)
        self._log_metrics(metrics.to_dict())
    
    def log_losses(self, **losses: float) -> None:
        """Log loss function values."""
        metrics = MetricGroup("loss")
        for name, value in losses.items():
            metrics.add(name, value)
        self._log_metrics(metrics.to_dict())
    
    def log_time(
        self,
        fps: float,
        total_timesteps: int,
        time_elapsed: float,
        iterations: Optional[int] = None
    ) -> None:
        """Log timing and performance metrics."""
        metrics = MetricGroup("time")
        metrics.add("fps", fps)
        metrics.add("total_timesteps", total_timesteps)
        metrics.add("time_elapsed", time_elapsed)
        if iterations is not None:
            metrics.add("iterations", iterations)
        self._log_metrics(metrics.to_dict())
    
    def _log_metrics(self, metrics: Dict[str, float]) -> None:
        """Internal method to dispatch metrics to all backends."""
        self._step += 1
        
        # Store in history
        for key, value in metrics.items():
            self._metrics_history[key].append((self._step, value))
        
        # TensorBoard
        if self.writer is not None:
            for key, value in metrics.items():
                self.writer.add_scalar(key, value, self._step)
        
        # Weights & Biases
        if self.use_wandb:
            import wandb
            wandb.log(metrics, step=self._step)
        
        # Console output
        if self.verbose > 0 and self._step % self.log_interval == 0:
            self._print_metrics(metrics)
    
    def _print_metrics(self, metrics: Dict[str, float]) -> None:
        """Pretty print metrics to console."""
        print(f"\n{'='*50}")
        print(f"Step: {self._step}")
        print(f"{'='*50}")
        for key, value in sorted(metrics.items()):
            print(f"  {key}: {value:.4f}")
```

### 4. Agent-Specific Implementation Tasks

- [ ] **SAC** (`tensoraerospace/agent/sac/`): Refactor to use BaseMetricsLogger
- [ ] **PPO** (`tensoraerospace/agent/ppo/`): Refactor to use BaseMetricsLogger
- [ ] **DDPG** (`tensoraerospace/agent/ddpg/`): Refactor to use BaseMetricsLogger
- [ ] **A3C** (`tensoraerospace/agent/a3c/`): Refactor to use BaseMetricsLogger
- [ ] **TD3** (if implemented): Refactor to use BaseMetricsLogger
- [ ] Add unit tests for metrics consistency across agents
- [ ] Update documentation with metrics reference guide

## Expected Outcomes

1. **Easy Algorithm Comparison**: Compare PPO vs SAC vs DDPG using identical metric names
2. **Simplified Integrations**: Single interface for TensorBoard, wandb, MLflow, etc.
3. **Better Documentation**: Clear reference for all available metrics
4. **Improved Maintainability**: New agents automatically follow established patterns
5. **Enhanced Debugging**: Consistent metrics make it easier to identify training issues

## Related Issues

- #176 — wandb integration (unified metrics will simplify this implementation)

## References

- [Stable-Baselines3 Logger](https://stable-baselines3.readthedocs.io/en/master/common/logger.html) — Reference implementation
- [OpenAI Spinning Up Logging](https://spinningup.openai.com/en/latest/) — Logging conventions

**Comment by mr8bit** [comment]
Implemented in #215. The PR introduces a unified TensorBoard metric schema and migrates all RL agents to it.

**What was delivered (matches the issue's checklist):**

✅ **Standardized metric groups** with stable prefixes — `rollout/`, `loss/`, `policy/`, `value/`, `diagnostics/`, `train/`, `eval/`, `weights/`, `grads/`. (The issue proposed `time/`; in practice everything that would have lived under `time/` — `total_timesteps`, `iterations`, `n_updates` — is covered by `rollout/total_steps`, `train/updates`, etc. `time/fps` and wall-clock are not logged from inside `MetricWriter` to keep training reproducibility cleaner; they can be added later if needed.)

✅ **Core metrics covered:**
- Rollout: `rollout/episode_reward`, `rollout/episode_length`, `rollout/total_steps` — plus `rollout/episode_reward_{median,p10,p90}` for PPO.
- Training: `train/lr`, `train/updates`, `train/replay_size`, `train/epsilon` (DQN), `train/per_beta` (DQN), etc.
- Loss: `loss/actor`, `loss/critic`, `loss/policy`, `loss/value`, `loss/entropy`, `loss/alpha` (SAC), `loss/q1`, `loss/q2` (SAC/DSAC), `loss/q` (DQN), `loss/discriminator` (GAIL), DHP variants for ADP.
- Diagnostics: `diagnostics/approx_kl`, `diagnostics/clip_fraction`, `diagnostics/explained_variance` (PPO), `diagnostics/terminated_count`, `diagnostics/truncated_count`.

✅ **Per-agent migration:** SAC, PPO, DDPG, A3C, A2C, A2C-NARX, DQN, DSAC, ADHDP, ADP — renamed in place. ET-DHP and GAIL had logging added from scratch.

✅ **Tests:** 3 unit-test files for the metrics package + 12 per-agent smoke tests verifying canonical tags actually appear in event files.

✅ **Documentation:** complete schema reference at `docs/en/guide/metrics.md`.

**Implementation differs from the issue's proposed `BaseMetricsLogger` design** in two ways:
1. Strict whitelist via constants in `schema.py` instead of a free-form `MetricGroup` (catches typos at write-time with a `ValueError`).
2. Mandatory minimum is enforced by `MetricWriter.assert_contract_satisfied()` at the end of `train()` — agents can no longer silently skip required metrics.

**Wandb integration (#176)** is not in this PR — it was kept out of scope to keep the migration focused. The unified schema lays the groundwork: a single sink switch (`MetricWriter` → `WandbMetricWriter`) is now a small follow-up.

Closing as completed.
